crates/verify/src/sourcify.rs

```diff
@@ -47,155 +64,317 @@ impl VerificationProvider for SourcifyVerificationProvider {
-            files.insert(import_entry, fs::read_to_string(&import)?);
-        }
+                // remove all incompatible settings
+                settings.sanitize(&context.compiler_version, SolcLanguage::Solidity);
+
+                input.settings = settings;
+
+                serde_json::to_value(&input).wrap_err("Failed to serialize standard json input")?
+            }
+            ContractLanguage::Vyper => {
+                let path = Path::new(&context.target_path);
+                let sources = Source::read_all_from(path, &["vy", "vyi"])?;
+                let input = VyperInput::new(
+                    sources,
+                    context.clone().compiler_settings.vyper,
+                    &context.compiler_version,
+                );
+
+                serde_json::to_value(&input).wrap_err("Failed to serialize vyper json input")?
+            }
+        };
+
+        let contract_identifier = format!(
+            "{}:{}",
+            context
+                .target_path
+                .strip_prefix(context.project.root())
+                .unwrap_or(context.target_path.as_path())
+                .display(),
+            context.target_name
+        );
+
+        let compiler_version = if matches!(lang, ContractLanguage::Vyper) {
+            context
+                .compiler_version
+                .clone()
+                .to_string()
+                .split('+')
+                .next()
+                .unwrap_or("0.0.0")
+                .to_string()
+        } else {
+            ensure_solc_build_metadata(context.compiler_version.clone()).await?.to_string()
+        };
 
         let req = SourcifyVerifyRequest {
-            address: args.address.to_string(),
-            chain: args.etherscan.chain.unwrap_or_default().id().to_string(),
-            files,
-            chosen_contract: None,
+            std_json_input,
+            compiler_version,
+            contract_identifier,
+            creation_transaction_hash: args.creation_transaction_hash.map(|h| h.to_string()),
         };
 
         Ok(req)
     }
 
-    fn process_sourcify_response(
-        &self,
-        response: Option<Vec<SourcifyResponseElement>>,
-    ) -> Result<()> {
-        let Some([response, ..]) = response.as_deref() else { return Ok(()) };
-        match response.status.as_str() {
-            "perfect" => {
-                if let Some(ts) = &response.storage_timestamp {
-                    sh_println!("Contract source code already verified. Storage Timestamp: {ts}")?;
+    async fn is_contract_verified(&self, args: &VerifyArgs) -> Result<bool> {
+        let chain_id = args.etherscan.chain.unwrap_or_default().id();
+        let base_url = args.verifier.verifier_url.as_deref().unwrap_or(SOURCIFY_URL);
+        let url = format!("{}v2/contract/{}/{}", base_url, chain_id, args.address);
+
+        match reqwest::get(&url).await {
+            Ok(response) => {
+                if response.status() == 200 {
```

Comment:
do we care about 200 only here or should we `if response.status().is_success()`?

Reply:
`is_success()` seems better :+1: